csrc/moe/moe_align_sum_kernels.cu

```diff
@@ -55,11 +55,11 @@ __global__ void moe_align_block_size_kernel(scalar_t* __restrict__ topk_ids,
   __syncthreads();
 
   // For each expert we accumulate the token counts from the different threads.
-  if (threadIdx.x < num_experts) {
-    tokens_cnts[index(num_experts, 0, threadIdx.x)] = 0;
+  for (int eid = threadIdx.x; eid < num_experts; eid += blockDim.x) {
+    tokens_cnts[index(num_experts, 0, eid)] = 0;
     for (int i = 1; i <= blockDim.x; ++i) {
```

Comment:
Does this have to be a nested loop?
Could a reduce operation be more efficient here?

Reply:
Each thread is doing its own reduce op on an independent data array. In this case warp reduce primitive does not apply.